### PR TITLE
Added Classes and Interfaces for API Response and State Typing

### PR DESCRIPTION
**Pull request by TBThomas56**
Fixes Issue #19 
Only to be merged after PR #24

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -10,21 +11,56 @@ import TableContainer from '@mui/material/TableContainer';
...
 
-type getMessageReturn = [string[],string[],string[],string[],number[],string[]]
+class MessageReturn {
+  timestamp: string[] = [];
+  host: string[] = [];
+  log_level_str: string[] = [];
+  log_message: string[] = [];
+  log_level: number[] = [];
+  app_name: string[] = [];
+}
```
Comment (callumforrester) [inline src/App.tsx:23]: You now have an excellent scope for simplifying the code:

What you're doing here is downloading a row-wise table from the server (which comes out as a list of `LogData`s) and then using a bunch of code to transpose into a column-wise table (a `MessageReturn` with lists of each category) and then applying the column-wise table to the DOM.

You can just cut out the middle man and have the state as a list of records, which would greatly simplify `getMessage()`. I would suggest something like this:

```suggestion
interface LogRecord {
  timestamp: string;
  host: string;
  log_level_str: string;
  log_message: string;
  log_level: number;
  app_name: string;
}
```

And then the state is

```python
const [LogResponse,setlogResponse] = useState<[LogRecord]>([]);
```

If you try that and refactor everything else around it I think the code will come out much simpler.

File: src/App.tsx [hunk 2]
```diff
@@ -10,21 +11,56 @@ import TableContainer from '@mui/material/TableContainer';
...
+interface LogData {
+  results: {
+      [key: string]: {
+        search_types: {
+          [key: string]: {
+            messages: Message[];
+          };
+      };
+  };
+}}
```
Comment (callumforrester) [inline src/App.tsx:46]: Good! I think these should go in their own source file to simplify things. I like the pato lays it out if you want to base it on something: https://github.com/DiamondLightSource/pato-frontend/blob/master/src/schema/interfaces.ts

File: src/App.tsx [hunk 3]
```diff
@@ -10,21 +11,56 @@ import TableContainer from '@mui/material/TableContainer';
...
 
 function App() {
-  const [time, setTime] = useState<string[]>([]);
-  const [host, setHost] = useState<string[]>([]);
-  const [debug, setDebug] = useState<string[]>([]);
-  const [logs, setLogs] = useState<string[]>([]);
-  const [log_lvl, setLog_lvl] = useState<number[]>([]);
-  const [app_name,setApp_name] = useState<string[]>([]);
+  // const [state, dispatch] = useReducer(reducer, new MessageReturn());
```
Comment (callumforrester) [inline src/App.tsx:49]: Can be deleted if unused

**Review by callumforrester — CHANGES_REQUESTED** [review]
Looking good...

File: src/App.tsx [hunk 4]
```diff
@@ -113,11 +108,12 @@ function App() {
       try {
         await readFile()
         .then((content) => {
-          username = content;
+
+          username.current = content;
```
Comment (callumforrester) [inline src/App.tsx:112]: Isn't `username` a string?

File: src/App.tsx [hunk 5]
```diff
@@ -168,37 +164,37 @@ function App() {
   );
 }
 
-function getMessage(logging: JSON): undefined | getMessageReturn {
-  const data = JSON.parse(JSON.stringify(logging));
+function getMessage(logging: JSON): LogRecord | undefined {
```
Comment (callumforrester) [inline src/App.tsx:167]: Related to https://github.com/DiamondLightSource/logpanel/pull/27#discussion_r1601243585

I think you should be able to come up with a class that maps directly to the API response, which hopefully makes this function much shorter.

File: src/App.tsx (hunk 4 above)
Comment (TBThomas56) [reply src/App.tsx:122]: https://stackoverflow.com/questions/57444154/why-need-useref-and-not-mutable-variable
found this is a reasoning behind why :)

**Pull request merged**